web/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql

```diff
@@ -0,0 +1,426 @@
+
+*/
+CREATE OR REPLACE PROCEDURE {{.output_schema}}.commit_with_metadata(
+                                                        META_TBL            VARCHAR,
+                                                        META_FROM_TBL       VARCHAR,
+                                                        STAGE_NXT           VARCHAR,
+                                                        STG_SRC_SCHEMA      VARCHAR,
+                                                        STG_SRC_TBL         VARCHAR,
+                                                        STG_TRG_SCHEMA      VARCHAR,
+                                                        STG_TRG_TBL         VARCHAR,
+                                                        STG_JOIN_KEY        VARCHAR,
+                                                        SKIP_DRV            VARCHAR,
+                                                        DRV_SRC_SCHEMA      VARCHAR,
+                                                        DRV_SRC_TBL         VARCHAR,
+                                                        DRV_TRG_SCHEMA      VARCHAR,
+                                                        DRV_TRG_TBL         VARCHAR,
+                                                        DRV_JOIN_KEY        VARCHAR,
+                                                        DRV_PARTITION_KEY   VARCHAR,
+                                                        DRV_UPSERT_LOOKBACK FLOAT)
+  RETURNS VARCHAR
+  LANGUAGE JAVASCRIPT
+  AS
+  $$
+
+  // Columns hardcoded for ease.
+  var meta_stmt = `INSERT INTO ` + META_TBL + `
+                     SELECT
+                       run_id,
+                       model_version,
+                       model,
+                       module,
+                       run_start_tstamp,
+                       CURRENT_TIMESTAMP::TIMESTAMP_NTZ AS run_end_tstamp,
+                       rows_this_run,
+                       distinct_key,
+                       distinct_key_count,
+                       time_key,
+                       min_time_key,
+                       max_time_key,
+                       duplicate_rows_removed,
+                       distinct_keys_removed
+                     FROM
+                       ` + META_FROM_TBL + `;`;
+
+  if (SKIP_DRV !== 'TRUE') {
+      var drv_trg_columns = list_cols(DRV_TRG_SCHEMA,DRV_TRG_TBL);
+      var drv_src_columns = list_cols(DRV_SRC_SCHEMA,DRV_SRC_TBL);
+      if (cols_check(drv_trg_columns, drv_src_columns) === false) {
+          throw Error("ERROR: column mismatch:" + DRV_TRG_TBL + "," + DRV_SRC_TBL);
+      }
+
+      var drv_trg_ref = `` + DRV_TRG_SCHEMA + `.` + DRV_TRG_TBL;
+      var drv_src_ref = `` + DRV_SRC_SCHEMA + `.` + DRV_SRC_TBL;
+      var drv_join_condition = `` + DRV_JOIN_KEY + ` IN
+                               (SELECT ` + DRV_JOIN_KEY + ` FROM ` + drv_src_ref + `)`;
+      var drv_lower_limit = `
```

Comment:
Can we execute the query and store the value in a var?

If so it'll be more efficient to just plug the value into the query than to have the queries nested.

Reply:
Yes, this should be possible. But why would that make it more efficient?

Reply:
Nested queries are inefficient in general in databases - because it has to create the query plan, execute the nested statement, and hold the output in memory for the remainder of the query (and also lookup that memory when executing the WHERE clause in this case).

In this case, it makes very little difference, because it's just one value. But it's still the more efficient structure to use.

Reply:
I am not an expert on the matter, but it is not a correlated subquery, so it will be calculated only once. The alternative of storing to a variable means we execute an extra query calling the JavaScript API one more time. For both cases there is a constant penalty, so it could be that the total time difference would be small. My assumption here on the memory lookups is that the subquery result will be cached. Of course i am open to it and willing to change it, just testing my assumptions.

Reply:
OK seems like a good reason to keep it as it is :)